Sources/StreamChat/Workers/ChannelListLinker.swift

```diff
@@ -0,0 +1,128 @@
+                notificationCenter: nc,
+                transform: { $0 as? ChannelVisibleEvent },
+                callback: { [weak self, databaseContainer] event in
+                    let context = databaseContainer.backgroundReadOnlyContext
+                    context.perform {
+                        guard let channel = try? context.channel(cid: event.cid)?.asModel() else { return }
+                        self?.linkChannelIfNeeded(channel)
+                    }
+                }
+            )
+        ]
+    }
+
+    private func isInChannelList(_ channel: ChatChannel, completion: @escaping (Bool) -> Void) {
+        let context = databaseContainer.backgroundReadOnlyContext
+        context.perform { [weak self] in
+            guard let self else { return }
+            if let (channelDTO, queryDTO) = context.getChannelWithQuery(cid: channel.cid, query: self.query) {
+                let isPresent = queryDTO.channels.contains(channelDTO)
+                completion(isPresent)
+            } else {
+                completion(false)
+            }
+        }
+    }
+    
+    /// Handles if a channel should be linked to the current query or not.
+    private func linkChannelIfNeeded(_ channel: ChatChannel) {
+        guard shouldChannelBelongToCurrentQuery(channel) else { return }
+        isInChannelList(channel) { [worker, query] exists in
+            guard !exists else { return }
+            worker.link(channel: channel, with: query) { error in
+                if let error = error {
+                    log.error(error)
+                    return
+                }
+                worker.startWatchingChannels(withIds: [channel.cid]) { error in
+                    guard let error = error else { return }
+                    log.warning(
+                        "Failed to start watching linked channel: \(channel.cid), error: \(error.localizedDescription)"
+                    )
+                }
+            }
+        }
+    }
+
+    /// Handles if a channel should be unlinked from the current query or not.
+    private func unlinkChannelIfNeeded(_ channel: ChatChannel) {
+        guard !shouldChannelBelongToCurrentQuery(channel) else { return }
+        isInChannelList(channel) { [worker, query] exists in
+            guard exists else { return }
+            worker.unlink(channel: channel, with: query)
+        }
+    }
+
+    /// Checks if the given channel should belong to the current query or not.
+    private func shouldChannelBelongToCurrentQuery(_ channel: ChatChannel) -> Bool {
+        if let filter = filter {
+            return filter(channel)
+        }
+
+        if clientConfig.isChannelAutomaticFilteringEnabled {
+            // When auto-filtering is enabled the channel will appear or not automatically if the
+            // query matches the DB Predicate. So here we default to saying it always belong to the current query.
+            return true
+        }
+
+        return false
```

Comment:
I went for the new implementation for this. The old was `deprecatedFallback() ?? true`.

Reply:
Yes, I guess we can remove the deprecatedFallback, but better if we also remove the old functions 👍